runtime/near-vm-runner/src/cache.rs

```diff
@@ -1,37 +1,88 @@
-#[cfg(not(feature = "no_cache"))]
-use cached::{cached_key, SizedCache};
-
 use crate::errors::IntoVMError;
 use crate::prepare;
+use near_primitives::borsh::BorshSerialize;
+use near_primitives::hash::CryptoHash;
 use near_vm_errors::VMError;
 use near_vm_logic::VMConfig;
+use std::fmt;
+use wasmer_runtime::{compiler_for_backend, Backend};
+use wasmer_runtime_core::cache::Artifact;
+use wasmer_runtime_core::load_cache_with;
 
-/// Cache size in number of cached modules to hold.
-#[cfg(not(feature = "no_cache"))]
-const CACHE_SIZE: usize = 128;
-// TODO: store a larger on-disk cache
-
-#[cfg(not(feature = "no_cache"))]
-cached_key! {
-    MODULES: SizedCache<(Vec<u8>, u64), Result<wasmer_runtime::Module, VMError>>
-        = SizedCache::with_size(CACHE_SIZE);
-    Key = {
-        (code_hash, config.non_crypto_hash())
-    };
-
-    fn compile_module(code_hash: Vec<u8>, code: &[u8], config: &VMConfig
-        ) -> Result<wasmer_runtime::Module, VMError> = {
-        let prepared_code = prepare::prepare_contract(code, config)?;
-        wasmer_runtime::compile(&prepared_code).map_err(|err| err.into_vm_error())
-    }
-}
-
-#[cfg(feature = "no_cache")]
 pub(crate) fn compile_module(
-    _code_hash: Vec<u8>,
+    _code_and_config_hash: Vec<u8>,
     code: &[u8],
     config: &VMConfig,
 ) -> Result<wasmer_runtime::Module, VMError> {
     let prepared_code = prepare::prepare_contract(code, config)?;
     wasmer_runtime::compile(&prepared_code).map_err(|err| err.into_vm_error())
 }
+
+pub fn get_hash(code: &[u8], config: &VMConfig) -> CryptoHash {
+    near_primitives::hash::hash(&[code, &config.non_crypto_hash().to_le_bytes()].concat())
```

Comment:
I'd recommend wrapping it into Enum and borshify it before hashing. This will allow us to update cached keys later. E.g.:
```rust
enum ContractCacheKey {
  Wasmer_0_17 {
    code_hash: CryptoHash,
    vm_config_non_crypto_hash: u64,
  }
}
```

Then use `near_primitives::hash::hash(key.try_to_vec().unwrap())`

Reply:
Interesting question is shall we hash the key, or use it as is?

Reply:
Probably we better should.